File: Proyecto-Factura-Contabilidad/src/app/servicios/configuracion.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment.development';
import { ApiResponse } from '../core/models/api-response';

export interface ConfiguracionDto {
  confId: number;
  ivaPorcentaje: number;
  establecimiento: string;
  puntoEmision: string;
  ultimoSecuencial: number;
  createdAt: string;
  updatedAt: string | null;
}

export interface ConfiguracionCrearDto {
  ivaPorcentaje: number;
  establecimiento: string;
  puntoEmision: string;
  ultimoSecuencial: number;
}

export interface ConfiguracionEditarDto extends ConfiguracionCrearDto {}

@Injectable({ providedIn: 'root' })
export class ConfiguracionService {
  private http = inject(HttpClient);
  private readonly API = environment.catalogoApiUrl;

  obtener(): Observable<ApiResponse<ConfiguracionDto>> {
    return this.http.get<ApiResponse<ConfiguracionDto>>(`${this.API}/api/Configuracion`);
  }

  crear(dto: ConfiguracionCrearDto): Observable<ApiResponse<ConfiguracionDto>> {
    return this.http.post<ApiResponse<ConfiguracionDto>>(`${this.API}/api/Configuracion`, dto);
  }

  editar(id: number, dto: ConfiguracionEditarDto): Observable<ApiResponse<any>> {
    return this.http.put<ApiResponse<any>>(`${this.API}/api/Configuracion/${id}`, dto);
  }
}

File: Microservicio.Catalogo.Api/Services/ConfiguracionService.cs
using Microsoft.EntityFrameworkCore;
using Microservicio.Catalogo.Api.Data;
using Microservicio.Catalogo.Api.DTOs.Configuracion;
using Microservicio.Catalogo.Api.Interfaces;
using Microservicio.Catalogo.Api.Models;

namespace Microservicio.Catalogo.Api.Services;

public class ConfiguracionService : IConfiguracionService
{
    private readonly CatalogoDbContext _context;

    public ConfiguracionService(CatalogoDbContext context)
    {
        _context = context;
    }

    public async Task<ConfiguracionDto?> ObtenerAsync()
    {
        var configuracion = await _context.Configuracions
            .AsNoTracking()
            .Where(c => c.DeletedAt == null)
            .OrderBy(c => c.ConfId)
            .FirstOrDefaultAsync();

        if (configuracion == null)
            return null;

        return new ConfiguracionDto
        {
            ConfId = configuracion.ConfId,
            IvaPorcentaje = configuracion.IvaPorcentaje,
            Establecimiento = configuracion.Establecimiento,
            PuntoEmision = configuracion.PuntoEmision,
            UltimoSecuencial = configuracion.UltimoSecuencial,
            CreatedAt = configuracion.CreatedAt,
            UpdatedAt = configuracion.UpdatedAt
        };
    }

    public async Task<ConfiguracionDto> CrearAsync(ConfiguracionCrearDto dto)
    {
        var existe = await _context.Configuracions
            .AnyAsync(c => c.DeletedAt == null);

        if (existe)
            throw new Exception("Ya existe una configuración registrada. Debe editar la existente.");

        var configuracion = new Configuracion
        {
            IvaPorcentaje = dto.IvaPorcentaje,
            Establecimiento = dto.Establecimiento.Trim(),
            PuntoEmision = dto.PuntoEmision.Trim(),
            UltimoSecuencial = dto.UltimoSecuencial,
            CreatedAt = DateTime.Now
        };

        _context.Configuracions.Add(configuracion);
        await _context.SaveChangesAsync();

        return new ConfiguracionDto
        {
            ConfId = configuracion.ConfId,
            IvaPorcentaje = configuracion.IvaPorcentaje,
            Establecimiento = configuracion.Establecimiento,
            PuntoEmision = configuracion.PuntoEmision,
            UltimoSecuencial = configuracion.UltimoSecuencial,
            CreatedAt = configuracion.CreatedAt,
            UpdatedAt = configuracion.UpdatedAt
        };
    }

    public async Task<bool> EditarAsync(int id, ConfiguracionEditarDto dto)
    {
        var configuracion = await _context.Configuracions
            .FirstOrDefaultAsync(c => c.ConfId == id && c.DeletedAt == null);

        if (configuracion == null)
            return false;

        configuracion.IvaPorcentaje = dto.IvaPorcentaje;
        configuracion.Establecimiento = dto.Establecimiento.Trim();
        configuracion.PuntoEmision = dto.PuntoEmision.Trim();
        configuracion.UltimoSecuencial = dto.UltimoSecuencial;
        configuracion.UpdatedAt = DateTime.Now;

        await _context.SaveChangesAsync();
        return true;
    }
}